all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+runtime validation.
+
+```rust
+trait SlotSupplier {
+  type SlotKind: SlotKind;
+  /// Blocks until a slot is available.
+  /// May also return an error if the backing user implementation encounters an error.
+  /// Returned errors will bubble up and cause the worker to shut down.
+  async fn reserve_slot(&self) -> Result<(), anyhow::Error>;
+
+  /// Tries to immediately reserve a slot, returning true if a slot is available.
+  /// May also return an error if the backing user implementation encounters an error.
+  /// Returned errors will bubble up and cause the worker to shut down.
+  fn try_reserve_slot(&self) -> Result<bool, anyhow::Error>;
+
+  /// Marks a slot as actually now being used. This is separate from reserving one because the pollers need to
+  /// reserve a slot before they have actually obtained work from server. Once that task is obtained (and validated)
+  /// then the slot can actually be used to work on the task.
+  /// 
+  /// Users' implementation of this can choose to emit metrics, or otherwise leverage the information provided by the
+  /// `info` parameter to be better able to make future decisions about whether a slot should be handed out.
+  fn mark_slot_used(&self, info: &Self::SlotKind::Info);
+
+  /// Frees a slot.
+  fn release_slot(&self, info: &Self::SlotKind::Info);
+  
+  /// If this implementation knows how many slots are available at any moment, it should return that here.
+  fn available_slots(&self) -> Option<usize>;
+}
+
+// These functions would be provided to the user as top level statics they can access to get information about how
+// all outstanding slots are currently being used. Ostensibly they could've stored this in their own implementation
+// too, but this is to make it convenient for them to answer questions like "How many slots are being used by 
+// activities of type X right now?". 
+//
+// These could potentially be provided as parameters to the trait methods - but at least in the case of the 
+// async/blocking reserve, the information may change over time, so it may be best to have these be explicit calls.
+pub fn used_workflow_slots_info() -> WorkflowSlotsInfo { /* ... */ }
+pub fn used_activity_slots_info() -> ActivitySlotsInfo { /* ... */ }
+pub fn used_local_activity_slots_info() -> LocalActivitySlotsInfo { /* ... */ }
+
+struct WorkflowSlotsInfo {
+  used_slots: Vec<WorkflowSlotInfo>,
+  /// Current size of the workflow cache.
+  num_cached_workflows: usize,
+  /// The limit on the size of the cache, if any. This is important for users to know as discussed below in the section
+  /// on workflow cache management.
+  max_cache_size: Option<usize>,
+  // ... Possibly also metric information
+}
+struct ActivitySlotsInfo {
+  used_slots: Vec<ActivitySlotInfo>,
+}
+struct LocalActivitySlotsInfo {
+  used_slots: Vec<LocalActivitySlotInfo>,
+}
+
+struct WorkflowSlotInfo {
+  workflow_type: String,
```

Comment:
Would we expose if these are eager workflows/activities here? I think it would be useful for some metrics